imitation/tabular_irl.py

```diff
@@ -0,0 +1,426 @@
+                # surely there's better way, though?
+                new_v = 0.0 * v + new_v
+                new_t.append(new_v)
+                idx_acc = new_idx_acc
+            new_params.append(new_t)
+        self._net_params = new_params
+
+    def get_params(self):
+        return self._flatten(self._net_params)
+
+
+class MLPRewardModel(JaxRewardModel):
+    """Simple MLP-based reward function with Jax/Stax."""
+    def __init__(self, obs_dim, hiddens, activation='Tanh', **kwargs):
+        assert activation in ['Tanh', 'Relu', 'Softplus'], \
+            "probably can't handle activation '%s'" % activation
+        self._hiddens = hiddens
+        self._activation = activation
+        super().__init__(obs_dim, **kwargs)
+
+    def make_stax_model(self):
+        act = getattr(jstax, self._activation)
+        layers = []
+        for h in self._hiddens:
+            layers.extend([jstax.Dense(h), act])
+        layers.extend([jstax.Dense(1), StaxSqueeze()])
+        return jstax.serial(*layers)
+
+
+def StaxSqueeze(axis=-1):
+    """Stax layer that collapses a single axis that has dimension 1."""
+    def init_fun(rng, input_shape):
+        ax = axis
+        if ax < 0:
+            ax = len(input_shape) + ax
+        assert ax < len(input_shape), \
+            "invalid axis %d for %d-dimensional tensor" \
+            % (axis, len(input_shape))
+        assert input_shape[ax] == 1, "axis %d is %d, not 1" \
+            % (axis, input_shape[ax])
+        output_shape = input_shape[:ax] + input_shape[ax + 1:]
+        return output_shape, ()
+
+    def apply_fun(params, inputs, **kwargs):
+        return jnp.squeeze(inputs, axis=axis)
+
+    return init_fun, apply_fun
+
+
+# ############################### #
+# ######### OPTIMISERS ########## #
+# ############################### #
+
+class Schedule(metaclass=abc.ABCMeta):
+    """Base class for learning rate schedules."""
+    @abc.abstractmethod
+    def __iter__(self):
+        """Yield an iterable of step sizes."""
+        pass
+
+
+class ConstantSchedule(Schedule):
+    """Constant step size schedule."""
+    def __init__(self, lr):
+        self.lr = lr
+
+    def __iter__(self):
+        while True:
+            yield self.lr
+
+
+class SqrtTSchedule(Schedule):
+    """1/sqrt(t) step size schedule."""
+    def __init__(self, init_lr):
+        self.init_lr = init_lr
+
+    def __iter__(self):
+        t = 1
+        while True:
+            yield self.init_lr / np.sqrt(t)
+            t += 1
+
+
+def get_schedule(lr_or_schedule):
+    """Turn a constant float/int or an actual Schedule into a canonical
+    Schedule instance."""
+    if isinstance(lr_or_schedule, Schedule):
+        return lr_or_schedule
+    if isinstance(lr_or_schedule, (float, int)):
+        return ConstantSchedule(lr_or_schedule)
+    raise TypeError("No idea how to make schedule out of '%s'" % lr_or_schedule)
+
+
+class Optimiser(metaclass=abc.ABCMeta):
```

Comment:
I can use implementations from Jax, I just didn't do that the first time around because I wanted to be 100% sure I knew what the optimisers were doing. I'll go fix that now.

Reply:
Also, the Jax optimiser library is a thing of beauty: https://github.com/google/jax/blob/97e16655155cc6e176e154ff932d7b2133e8cf1d/jax/experimental/optimizers.py#L180-L369

Compare that to the source of `tf.train` :stuck_out_tongue_winking_eye: